### PR TITLE
server: Add status vars for NotAfter/NotBefore of TLS certs

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

This adds `Ssl_server_not_before` and `Ssl_server_not_after` as status
variables. This helps to identify the expiry date of the certificate
that was loaded into the server.

Reading the cert on disk may not be sufficient if `ALTER INSTANCE RELOAD
TLS` was not (yet) executed.

This improves MySQL compatibility.

```
tidb> SHOW GLOBAL STATUS LIKE 'Ssl\_server\_%';
+-----------------------+-------------------------+
| Variable_name         | Value                   |
+-----------------------+-------------------------+
| Ssl_server_not_after  | Oct 6 15:06:29 2021 UTC |
| Ssl_server_not_before | Jul 8 15:06:29 2021 UTC |
+-----------------------+-------------------------+
2 rows in set (0.0010 sec)
```

### What is changed and how it works?

This adds status variables to `Server`.

There are some other `Ssl_*` variables defined in `sessionctx/variable/statusvar.go` but those don't need or have access to `tlsConfig`.

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note 

- Details of the TLS certificate of the server are now available in the `Ssl_server_not_before` and `Ssl_server_not_after` status variables

### CODE REVIEW

**Comment by morgo** [comment]
Getting linter errors:
```
[2021-07-12T07:01:57.208Z] server/stat.go:30:19: *github.com/pingcap/tidb/sessionctx/variable.StatusVal composite literal uses unkeyed fields

[2021-07-12T07:01:57.208Z] server/stat.go:31:19: *github.com/pingcap/tidb/sessionctx/variable.StatusVal composite literal uses unkeyed fields
```

**Comment by morgo** [comment]
@dveeden Do not_after and not_before have clear meanings in TLS-land?

I am not an expert on TLS, but the meaning doesn't jump out at me so I am wondering if there could be something clearer.

**Comment by dveeden** [comment]
> @dveeden Do not_after and not_before have clear meanings in TLS-land?
> 
> I am not an expert on TLS, but the meaning doesn't jump out at me so I am wondering if there could be something clearer.

This is how it looks in OpenSSL output:
```
$ openssl x509 -in tls/cert.pem -noout -text | grep -A2 'Validity'
        Validity
            Not Before: Jul  8 15:06:29 2021 GMT
            Not After : Oct  6 15:06:29 2021 GMT
```

In [x509.Certificate](https://pkg.go.dev/crypto/x509#Certificate) they are called `NotBefore`, `NotAfter`.

These names can also be found in [RFC 5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5) as `notBefore` and `notAfter`.

I used these names to match [the names in MySQL](https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Ssl_server_not_after)

A bigger issue with the name is the mixing of "SSL" and "TLS". "SSL" is old and not used by MySQL or TiDB. And "TLS" is new. So the "Ssl_*" variables were poorly named and never fixed. However I think we should stick with them in TiDB to be compatible with MySQL and because that is what people and applications expect

**Comment by morgo** [comment]
Makes sense, thanks for clarifying :-)

(I wish MySQL had named it something like `Ssl_server_validity_not_after`, but I agree we should not change it.)

**Comment by dveeden** [comment]
> (I wish MySQL had named it something like `Ssl_server_validity_not_after`, but I agree we should not change it.)

`Tls_server_validity_not_after` ... or `Tls.server.validity.not_after` ... but yes let's not change that

File: server/server.go [hunk 1]
```diff
@@ -268,6 +268,9 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 
 	// Init rand seed for randomBuf()
 	rand.Seed(time.Now().UTC().UnixNano())
+
+	variable.RegisterStatistics(s)
```
Comment (mjonss) [inline server/server.go:272]: Is the whole server struct needed to be sent for registration or is it just to have a reference to the needed part of the config/tls?

**Review by mjonss — COMMENTED** [review]
Would it be possible to add a cert and key file with a very long valid time and use it for a unit test?

**Review by morgo — APPROVED** [review]

**Comment by dveeden** [comment]
> Would it be possible to add a cert and key file with a very long valid time and use it for a unit test?

Even better: There is [generateCert](https://github.com/pingcap/tidb/blob/350cbd1f69c47df1479f98467d214540d834384e/server/tidb_test.go#L753) that is already used to generate test certs. I just added testing for added status vars.

File: server/stat.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	tlsConfig := s.getTLSConfig()
+	if tlsConfig != nil {
+		if len(tlsConfig.Certificates) == 1 {
+			pc, err := x509.ParseCertificate(tlsConfig.Certificates[0].Certificate[0])
+			if err != nil {
+				logutil.BgLogger().Error("Failed to parse TLS certficates to get server status", zap.Error(err))
+			} else {
+				m[serverNotAfter] = pc.NotAfter.Format("Jan 2 15:04:05 2006 MST")
```
Comment (mjonss) [inline server/stat.go:54]:
```suggestion
				m[serverNotAfter] = pc.NotAfter.Format("Jan _2 15:04:05 2006 MST")
```
I just checked MySQL 8.0.25 and it shows:
```
May  5 08:48:56 2021 GMT
```
I.e. with an extra space between Month and Day, so I wonder if it should be space-padded?

MySQL seems to use the openssl function `ASN1_TIME_print` which resolves to `ossl_asn1_time_print_ex` with the format:
```c
527             return BIO_printf(bp, "%s %2d %02d:%02d:%02d%.*s %d%s",
528                           _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
529                           stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
530                           (gmt ? " GMT" : "")) > 0;
```

**Review by mjonss — APPROVED** [review]
LGTM (but I'm not approved as reviewer yet).

Just one minor thing about the date format.

File: server/stat.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	tlsConfig := s.getTLSConfig()
+	if tlsConfig != nil {
+		if len(tlsConfig.Certificates) == 1 {
+			pc, err := x509.ParseCertificate(tlsConfig.Certificates[0].Certificate[0])
+			if err != nil {
+				logutil.BgLogger().Error("Failed to parse TLS certficates to get server status", zap.Error(err))
+			} else {
+				m[serverNotAfter] = pc.NotAfter.Format("Jan _2 15:04:05 2006 MST")
```
Comment (tiancaiamao) [inline server/stat.go:54]: Is `_2` a typo, should it be 2?

**Comment by tiancaiamao** [comment]
Rest LGTM

File: server/stat.go (hunk 3 above)
Comment (dveeden) [reply server/stat.go:54]: That is not a typo. That is to match the OpenSSL formatting as explained by @mjonss in https://github.com/pingcap/tidb/pull/26132#discussion_r672962217

A simple example:
https://play.golang.org/p/a5BVMAKuixs

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**